src/js/angular/clustermanagement/directives.js

```diff
@@ -58,7 +91,48 @@ clusterManagementDirectives.directive('clusterGraphicalView', ['$window', 'Local
                 let nodesGroup;
                 let linksGroup;
 
+                function updateLabels() {
+                    CDS.updateClusterZoneLabels(hasCluster, clusterZone, translationsMap);
+                    Object.keys(idTranslationKeyMap).forEach((elementId) => {
+                        d3.select(`#${elementId}`).text(getLabelFor(elementId));
+                    });
+                    if (legendNodesGroup) {
+                        legendNodesGroup.remove();
+                    }
+                    legendNodesGroup = svg.append('g').attr('id', 'legend-nodes-group').classed('hidden', !scope.showLegend);
+                    createClusterLegend();
+                }
+
+                function translateLegendLinksAndNodes(legendNodes, legendLinks) {
+                    const maxLabelLength = 13;
+
+                    legendNodes.forEach((node) => {
+                        node.fullLabel = translationsMap[node.customText];
+                        node.hostname = node.fullLabel;
+                        if (node.fullLabel && node.fullLabel.length > maxLabelLength) {
+                            const end = maxLabelLength + 3 > node.fullLabel.length ? node.fullLabel.length : maxLabelLength;
+                            node.hostname = node.fullLabel.substring(0, end - 4) + '...';
+                            node.shortened = true;
+                        }
+                    });
+                    legendLinks.forEach((link) => {
+                        link.linkTypeText = translationsMap[link.linkTypeKey];
+                        link.fullLabel = translationsMap[link.linkTypeKey];
+                        link.linkTypeText = link.fullLabel;
+                        if (link.fullLabel && link.fullLabel.length > maxLabelLength) {
```

Comment:
Not sure how precise this approach is, but does it consider the letter dimensions?

Reply:
Refactored to use a `foreignObject` with a div inside. Set text overflow with css.